jina/optimize/__init__.py

```diff
@@ -0,0 +1,231 @@
+import os
+from itertools import tee
+import json
+from pathlib import Path
+import shutil
+from ruamel.yaml import YAML
+
+from jina.flow import Flow
+from jina.helper import colored
+from jina.logging import default_logger as logger
+from type import IntegerParameter, load_optimization_parameters
+
+
+class FlowRunner:
+    def __init__(
+        self,
+        index_document_generator,
```

Comment:
Why would I need to pass query_document_generator if I do not want to evaluate a query flow? It feels that this class is very specialized in a single way of evaluating.

Reply:
I will keep them `None`

Reply:
I still do not think we should own a generator.

- Is it a generator or a generator function? if it is a generator it will only be available once, so owning something that will not be useful during the instance lifetime is not good

- Is the generator something that the user needs to know when instantiating the FlowRunner?

Reply:
It is a generator. `tee` is used to make a copy of the generator so that we can rerun generators. We will think of moving it to run methods